File: Source/WickedSick.Server.XamlParser/Elements/Data/Binding.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Globalization;

namespace WickedSick.Server.XamlParser.Elements.Data
{
    [Element("Fayde.Data")]
    public enum BindingMode
    {
        OneWay,
        OneTime,
        TwoWay
    }

    [Element("Fayde.Data")]
    public enum UpdateSourceTrigger
    {
        Default,
        Explicit
    }

    [Element("Fayde", "BindingMarkup")]
    public class Binding : IJsonConvertible
    {
        public object FallbackValue { get; set; }
        public string Path { get; set; }
        public BindingMode Mode { get; set; }
        public object Source { get; set; }
        public bool BindsDirectlyToSource { get; set; }
        public string StringFormat { get; set; }
        public object TargetNullValue { get; set; }
        public IValueConverter Converter { get; set; }
        public CultureInfo ConverterCulture { get; set; }
        public object ConverterParameter { get; set; }
        public bool NotifyOnValidationError { get; set; }
        public bool ValidatesOnExceptions { get; set; }
        public bool ValidatesOnDataErrors { get; set; }
        public bool ValidatesOnNotifyDataErrors { get; set; }
        public RelativeSource RelativeSource { get; set; }
        public string ElementName { get; set; }
        public UpdateSourceTrigger UpdateSourceTrigger { get; set; }

        public Binding()
        {
        }

        public string ToJson(int tabIndents, IJsonOutputModifiers outputMods)
        {
            StringBuilder sb = new StringBuilder();
            sb.Append("new ");
            sb.Append(ElementAttribute.GetFullNullstoneType(GetType(), outputMods));
            sb.Append("({ ");
            sb.Append(string.Join(", ", GetPropertiesJson(tabIndents, outputMods)));
            sb.Append(" })");
            return sb.ToString();
        }

        private IEnumerable<string> GetPropertiesJson(int tabIndents, IJsonOutputModifiers outputMods)
        {
            if (Mode != BindingMode.OneWay)
                yield return string.Format("Mode: {0}.{1}", ElementAttribute.GetFullNullstoneType(typeof(BindingMode), outputMods), Mode);
            if (Path != null)
                yield return string.Format("Path: \"{0}\"", Path);
            if (FallbackValue != null)
            {
                var ijc = FallbackValue as IJsonConvertible;
                if (ijc != null)
                    yield return string.Format("FallbackValue: {0}", ijc.ToJson(tabIndents, outputMods));
                else if (FallbackValue is string)
                    yield return string.Format("FallbackValue: \"{0}\"", FallbackValue);
                else
                    yield return string.Format("FallbackValue: {0}", FallbackValue);
            }
            if (StringFormat != null)
                yield return string.Format("StringFormat: \"{0}\"", StringFormat);
            if (ElementName != null)
                yield return string.Format("ElementName: \"{0}\"", ElementName);
        }
    }
}

File: src/Data/Binding.ts
module Fayde.Data {
    import convert = nullstone.convertAnyToType;

    export var WarnBrokenPath = false;

    export class Binding implements nullstone.markup.IMarkupExtension, ICloneable {
        StringFormat: string;
        FallbackValue: any;
        TargetNullValue: any;
        BindsDirectlyToSource: boolean;
        Converter: IValueConverter;
        ConverterParameter: any;
        ConverterCulture: any;
        ElementName: string;
        Mode: BindingMode;
        NotifyOnValidationError: boolean;
        RelativeSource: RelativeSource;
        Path: Data.PropertyPath;
        Source: any;
        UpdateSourceTrigger: UpdateSourceTrigger;
        ValidatesOnExceptions: boolean;
        ValidatesOnDataErrors: boolean;
        ValidatesOnNotifyDataErrors: boolean;

        constructor ();
        constructor (path: string);
        constructor (path: Data.PropertyPath);
        constructor (binding: Binding);
        constructor (obj?: any) {
            if (obj instanceof Binding) {
                var binding = <Binding>obj;
                this.StringFormat = binding.StringFormat;
                this.FallbackValue = binding.FallbackValue;
                this.TargetNullValue = binding.TargetNullValue;
                this.BindsDirectlyToSource = binding.BindsDirectlyToSource;
                this.Converter = binding.Converter;
                this.ConverterParameter = binding.ConverterParameter;
                this.ConverterCulture = binding.ConverterCulture;
                this.ElementName = binding.ElementName;
                this.Mode = binding.Mode;
                this.NotifyOnValidationError = binding.NotifyOnValidationError;
                this.RelativeSource = binding.RelativeSource ? binding.RelativeSource.Clone() : null;
                this.Path = binding.Path;
                this.Source = binding.Source;
                this.UpdateSourceTrigger = binding.UpdateSourceTrigger;
                this.ValidatesOnExceptions = binding.ValidatesOnExceptions;
                this.ValidatesOnDataErrors = binding.ValidatesOnDataErrors;
                this.ValidatesOnNotifyDataErrors = binding.ValidatesOnNotifyDataErrors;
            } else if (typeof obj === "string") {
                this.Path = new Data.PropertyPath(<string>obj);
            } else if (obj instanceof Data.PropertyPath) {
                this.Path = obj;
            } else {
                this.Path = new Data.PropertyPath("");
            }
        }

        init (val: string) {
            this.Path = new Data.PropertyPath(val);
        }

        transmute (os: any[]): any {
            this.$$coerce();
            Object.freeze(this);
            return new Data.BindingExpression(this);
        }

        private $$coerce () {
            this.StringFormat = this.StringFormat ? this.StringFormat.toString() : undefined;
            this.BindsDirectlyToSource = convert(this.BindsDirectlyToSource, Boolean);
            this.Mode = Enum.fromAny(BindingMode, this.Mode);
            this.NotifyOnValidationError = convert(this.NotifyOnValidationError, Boolean);
            this.Path = convert(this.Path, Data.PropertyPath);
            this.UpdateSourceTrigger = Enum.fromAny(UpdateSourceTrigger, this.UpdateSourceTrigger);
            this.ValidatesOnExceptions = convert(this.ValidatesOnExceptions, Boolean);
            this.ValidatesOnDataErrors = convert(this.ValidatesOnDataErrors, Boolean);
            this.ValidatesOnNotifyDataErrors = convert(this.ValidatesOnNotifyDataErrors, Boolean);
        }

        Clone () {
            return new Binding(this);
        }
    }
    Fayde.CoreLibrary.add(Binding);
}